File: scripts/k6/load.js
// k6 load test for the BNI Natcon 2026 API.
//
// Run against a FRESH database (setup seeds its own members):
//
//   createdb -O natcon natcon_k6
//   ADDR=:8084 DATABASE_URL=postgres://natcon:natcon@localhost:5432/natcon_k6?sslmode=disable \
//     go run ./backend/cmd/api &
//   BASE=http://localhost:8084 k6 run scripts/k6/load.js
//
// Scenarios (run concurrently, ~3 minutes total):
//   browse — attendee journey (me / tenants / seminars / networking),
//            ramping 0 → 60 VUs.
//   scan   — booth scanner firing POST /scans at a constant 20 rps.
//   admin  — committee dashboard polling (overview/tenants/seminars/activity).
//
// The login endpoint is rate-limited per IP, so setup() logs in with a
// unique X-Forwarded-For per request (RealIP middleware honors it).

import http from 'k6/http'
import { check, sleep } from 'k6'
import { Rate, Trend } from 'k6/metrics'

const BASE = __ENV.BASE || 'http://localhost:8084'
const PASSWORD = __ENV.SEED_PASSWORD || 'natcon2026'
const MEMBER_LOGINS = 20
const SEED_MEMBERS = 200

const scanDuration = new Trend('scan_duration', true)
const scanErrors = new Rate('scan_errors')

export const options = {
  scenarios: {
    browse: {
      executor: 'ramping-vus',
      exec: 'memberJourney',
      startVUs: 0,
      stages: [
        { duration: '30s', target: 20 },
        { duration: '60s', target: 60 },
        { duration: '45s', target: 60 },
        { duration: '15s', target: 0 },
      ],
    },
    scan: {
      executor: 'constant-arrival-rate',
      exec: 'boothScan',
      rate: 20,
      timeUnit: '1s',
      duration: '2m30s',
      preAllocatedVUs: 30,
      maxVUs: 60,
    },
    admin: {
      executor: 'constant-vus',
      exec: 'adminDashboard',
      vus: 3,
      duration: '2m30s',
    },
  },
  thresholds: {
    http_req_failed: ['rate<0.01'],
    http_req_duration: ['p(95)<300', 'p(99)<800'],
    'http_req_duration{scenario:browse}': ['p(95)<300'],
    'http_req_duration{scenario:scan}': ['p(95)<300'],
    'http_req_duration{scenario:admin}': ['p(95)<400'],
    checks: ['rate>0.99'],
    scan_errors: ['rate<0.01'],
  },
}

const jsonHeaders = { 'Content-Type': 'application/json' }

function login(email, password, xff) {
  const res = http.post(
    `${BASE}/api/v1/auth/login`,
    JSON.stringify({ email, password }),
    { headers: { ...jsonHeaders, 'X-Forwarded-For': xff } }
  )
  if (res.status !== 200) {
    throw new Error(`login ${email} failed: ${res.status} ${res.body}`)
  }
  return res.json('token')
}

export function setup() {
  const adminTok = login('admin@natcon.id', PASSWORD, '10.66.0.1')
  const tenantTok = login('booth-a03@natcon.id', PASSWORD, '10.66.0.2')

  // Seed load-test members via the upsert import (idempotent on re-runs).
  // Generated password = chapter+firstname slug -> "chapterk6k6" for all.
  for (let start = 0; start < SEED_MEMBERS; start += 100) {
    const rows = []
    for (let i = start; i < Math.min(start + 100, SEED_MEMBERS); i++) {
      const n = String(i).padStart(4, '0')
      rows.push({
        name: `K6 Member ${n}`,
        email: `k6-${n}@natcon.id`,
        chapter: 'Chapter K6',
        phone: `+62899${n}000`,
      })
    }
    const res = http.post(`${BASE}/api/v1/admin/members/bulk`, JSON.stringify({ members: rows }), {
      headers: { ...jsonHeaders, Authorization: `Bearer ${adminTok}` },
    })
    if (res.status !== 200) {
      throw new Error(`bulk seed failed: ${res.status} ${res.body}`)
    }
  }

  const list = http.get(`${BASE}/api/v1/admin/members?q=k6-&limit=1000`, {
    headers: { Authorization: `Bearer ${adminTok}` },
  })
  const memberCodes = list.json('members').map((m) => m.member_code)

  const memberToks = []
  for (let i = 0; i < MEMBER_LOGINS; i++) {
    const n = String(i).padStart(4, '0')
    memberToks.push(login(`k6-${n}@natcon.id`, 'chapterk6k6', `10.66.1.${i + 1}`))
  }

  return { adminTok, tenantTok, memberToks, memberCodes }
}

const authGet = (path, token) =>
  http.get(`${BASE}${path}`, { headers: { Authorization: `Bearer ${token}` } })

export function memberJourney(data) {
  const token = data.memberToks[__VU % data.memberToks.length]

  let res = authGet('/api/v1/me', token)
  check(res, { 'me 200': (r) => r.status === 200 })
  sleep(Math.random() * 0.7 + 0.3)

  res = authGet('/api/v1/tenants', token)
  check(res, {
    'tenants 200': (r) => r.status === 200,
    'tenants has sponsors first': (r) => r.json('tenants.0.kind') === 'sponsor',
  })
  sleep(Math.random() * 0.7 + 0.3)

  res = authGet('/api/v1/seminars', token)
  check(res, { 'seminars 200': (r) => r.status === 200 })
  sleep(Math.random() * 0.7 + 0.3)

  res = authGet('/api/v1/networking', token)
  check(res, { 'networking 200': (r) => r.status === 200 })
  sleep(Math.random() * 1.0 + 0.5)
}

export function boothScan(data) {
  const code = data.memberCodes[Math.floor(Math.random() * data.memberCodes.length)]
  const res = http.post(`${BASE}/api/v1/scans`, JSON.stringify({ member_code: code }), {
    headers: { ...jsonHeaders, Authorization: `Bearer ${data.tenantTok}` },
    tags: { name: 'POST /scans' },
  })
  scanDuration.add(res.timings.duration)
  scanErrors.add(res.status !== 200)
  check(res, {
    'scan 200': (r) => r.status === 200,
    'scan identifies member': (r) => String(r.json('member_name') || '').startsWith('K6 Member'),
  })
}

export function adminDashboard(data) {
  const t = data.adminTok
  check(authGet('/api/v1/admin/overview', t), { 'overview 200': (r) => r.status === 200 })
  check(authGet('/api/v1/admin/tenants', t), { 'ranking 200': (r) => r.status === 200 })
  check(authGet('/api/v1/admin/seminars', t), { 'fill 200': (r) => r.status === 200 })
  check(authGet('/api/v1/admin/activity?limit=15', t), { 'activity 200': (r) => r.status === 200 })
  sleep(1)
}

File: scripts/load.py
#!/usr/bin/env python3
"""Event-scale load test: the whole venue on the API at once.

769 attendees hold tickets; this drives N of them (default 700) through the
day's four rushes against a FRESH database seeded with the real event data:

  1. The morning sign-in storm — real /auth/login calls, bcrypt and all,
     each phone on its own IP (X-Forwarded-For), shared-email tickets
     resolved through the account chooser.
  2. The browse storm — every signed-in phone loads /me, /tenants,
     /seminars and /rundown at once.
  3. The class rush — everyone tries to grab a seat the moment
     registration opens; the four classes hold 60 each, so exactly 240
     may win and nobody may be oversold.
  4. The expo — booths scan the crowd; and networking check-in fills
     88 tables of 8 with zero oversell.

Correctness is asserted; latency is reported (p50/p95/p99). Run it before
the event on hardware shaped like production:

    createdb natcon_load
    ADDR=:8099 DATABASE_URL=postgres://...natcon_load... go run ./backend/cmd/api &
    ulimit -n 4096   # 700 keep-alive sockets need file descriptors
    BASE=http://localhost:8099 N=700 python3 scripts/load.py

Stdlib only; exits non-zero on any failed assertion.
"""

import http.client
import json
import os
import re
import statistics
import sys
import threading
import time
import unicodedata
from concurrent.futures import ThreadPoolExecutor
from urllib.parse import urlparse

BASE = os.environ.get("BASE", "http://localhost:8099")
N = int(os.environ.get("N", "700"))
PASSWORD = os.environ.get("SEED_PASSWORD", "natcon2026")
parsed = urlparse(BASE)
HOST, PORT = parsed.hostname, parsed.port or 80

passed = failed = 0
lock = threading.Lock()


def check(name, cond, detail=""):
    global passed, failed
    if cond:
        passed += 1
        print(f"  ok   {name}")
    else:
        failed += 1
        print(f"  FAIL {name} {detail}")


class Client:
    def __init__(self):
        self.conn = http.client.HTTPConnection(HOST, PORT, timeout=60)

    def req(self, method, path, token=None, body=None, xff=None):
        headers = {}
        payload = None
        if body is not None:
            payload = json.dumps(body)
            headers["Content-Type"] = "application/json"
        if token:
            headers["Authorization"] = f"Bearer {token}"
        if xff:
            headers["X-Forwarded-For"] = xff
        for attempt in (1, 2):
            try:
                self.conn.request(method, path, body=payload, headers=headers)
                resp = self.conn.getresponse()
                data = resp.read()
                break
            except Exception:
                self.conn.close()
                self.conn = http.client.HTTPConnection(HOST, PORT, timeout=60)
                if attempt == 2:
                    raise
        out = None
        if data:
            try:
                out = json.loads(data)
            except json.JSONDecodeError:
                out = None
        return resp.status, out


def pct(lat, p):
    return statistics.quantiles(lat, n=100)[p - 1] if len(lat) > 1 else lat[0]


def report(name, lat, wall, statuses):
    ok = sum(v for k, v in statuses.items() if 200 <= k < 300)
    print(f"  {name}: {len(lat)} reqs in {wall:.1f}s ({len(lat)/wall:.0f} rps) · "
          f"p50 {pct(lat, 50)*1000:.0f}ms · p95 {pct(lat, 95)*1000:.0f}ms · "
          f"p99 {pct(lat, 99)*1000:.0f}ms · statuses {dict(sorted(statuses.items()))} · 2xx {ok}")


def first_password(chapter, name):
    """Mirror of scripts/attendees_migration.py — what the ticket says."""
    first = name.split()[0] if name.split() else ""
    raw = unicodedata.normalize("NFKD", f"{chapter}{first}")
    return re.sub(r"\s+", "", raw).lower()


def run_pool(items, fn, workers=None):
    lat, statuses, results = [], {}, []
    t0 = time.time()
    with ThreadPoolExecutor(max_workers=workers or len(items)) as ex:
        for status, elapsed, extra in ex.map(fn, items):
            lat.append(elapsed)
            statuses[status] = statuses.get(status, 0) + 1
            results.append((status, extra))
    return lat, statuses, results, time.time() - t0


# ---------------------------------------------------------------- setup
boot = Client()
status, body = boot.req("POST", "/api/v1/auth/login",
                        body={"email": "admin@natcon.id", "password": PASSWORD})
assert status == 200, f"admin login failed: {status} {body}"
admin_tok = body["token"]

members = []
page = 1
while True:
    status, body = boot.req("GET", f"/api/v1/admin/members?limit=1000&page={page}", token=admin_tok)
    assert status == 200
    members += body["members"]
    if len(members) >= body["total"] or not body["members"]:
        break
    page += 1
crowd = members[:N]
print(f"== Setup == {len(members)} attendees seeded, driving {len(crowd)}")

status, body = boot.req("GET", "/api/v1/admin/seminars", token=admin_tok)
seminars = body["seminars"]
capacity_total = sum(s["capacity"] for s in seminars)

TABLES = -(-N // 8) + 1  # enough 8-seat tables for everyone, plus one spare
status, body = boot.req("POST", "/api/v1/admin/tables/generate", token=admin_tok,
                        body={"count": TABLES, "hall": "Hall Load", "capacity": 8})
assert status in (200, 201), f"tables: {status} {body}"

# ------------------------------------------- 1. the morning sign-in storm
print(f"\n== 1. Sign-in storm ({len(crowd)} phones at once) ==")

def sign_in(im):
    i, m = im
    c = Client()
    pw = first_password(m["chapter"], m["name"])
    t0 = time.time()
    status, body = c.req("POST", "/api/v1/auth/login",
                         body={"email": m["email"], "password": pw},
                         xff=f"10.{i // 250}.{(i // 50) % 5}.{i % 50 + 1}")
    if status == 200 and body and "accounts" in body:
        status, body = c.req("POST", "/api/v1/auth/login/select",
                             body={"choice_token": body["choice_token"], "user_id": m["id"]},
                             xff=f"10.{i // 250}.{(i // 50) % 5}.{i % 50 + 1}")
    elapsed = time.time() - t0
    tok = body.get("token") if status == 200 and body else None
    return status, elapsed, (m, tok, c)

lat, statuses, results, wall = run_pool(list(enumerate(crowd)), sign_in)
report("login", lat, wall, statuses)
tokens = [(m, tok, c) for _, (m, tok, c) in results if tok]
check(f"all {len(crowd)} attendees signed in", len(tokens) == len(crowd),
      f"got {len(tokens)}; statuses {statuses}")

# ------------------------------------------------------- 2. browse storm
print(f"\n== 2. Browse storm ({len(tokens)} × 4 endpoints) ==")

def browse(entry):
    m, tok, c = entry
    t0 = time.time()
    worst = 200
    for path in ("/api/v1/me", "/api/v1/tenants", "/api/v1/seminars", "/api/v1/rundown"):
        s, _ = c.req("GET", path, token=tok)
        worst = max(worst, s)
    return worst, time.time() - t0, None

lat, statuses, _, wall = run_pool(tokens, browse)
report("browse (4 GETs each)", lat, wall, statuses)
check("every phone loaded every screen", statuses.get(200, 0) == len(tokens), f"{statuses}")

# --------------------------------------------------------- 3. class rush
print(f"\n== 3. Class rush ({len(tokens)} people, {capacity_total} seats) ==")

def grab_seat(ie):
    i, (m, tok, c) = ie
    sem = seminars[i % len(seminars)]
    t0 = time.time()
    s, _ = c.req("POST", f"/api/v1/seminars/{sem['id']}/register", token=tok)
    return s, time.time() - t0, sem["id"]

lat, statuses, results, wall = run_pool(list(enumerate(tokens)), grab_seat)
report("register", lat, wall, statuses)
won = statuses.get(200, 0) + statuses.get(201, 0)
check(f"exactly {capacity_total} seats won, the rest turned away cleanly",
      won == capacity_total
      and statuses.get(409, 0) == len(tokens) - capacity_total
      and not any(k >= 500 for k in statuses), f"{statuses}")

status, body = boot.req("GET", "/api/v1/admin/seminars", token=admin_tok)
check("no class oversold",
      all(s["seats_taken"] <= s["capacity"] for s in body["seminars"])
      and sum(s["seats_taken"] for s in body["seminars"]) == capacity_total,
      f'{[(s["room"], s["seats_taken"], s["capacity"]) for s in body["seminars"]]}')

# ------------------------------------------------- 4a. booths scan the crowd
print(f"\n== 4a. Expo scan storm ({len(tokens)} scans across the booths) ==")

status, body = boot.req("GET", "/api/v1/admin/tenants", token=admin_tok)
booths = [t for t in body["tenants"] if t["kind"] == "booth"]

def booth_token(t):
    c = Client()
    from_name = "".join(ch for ch in (t["name"] + t["booth"]).lower() if ch.isalnum() and ch.isascii())
    s, b = c.req("POST", "/api/v1/auth/login",
                 body={"email": f"booth-{t['booth'].split(' ')[0].replace('-', '').lower()}@natcon.id",
                       "password": from_name},
                 xff=f"10.9.{booths.index(t) // 50}.{booths.index(t) % 50 + 1}")
    return (s, b.get("token") if b else None, c)

booth_sessions = [booth_token(t) for t in booths]
booth_sessions = [(tok, c) for s, tok, c in booth_sessions if s == 200 and tok]
check(f"every booth scanner signed in ({len(booths)})",
      len(booth_sessions) == len(booths), f"got {len(booth_sessions)}")

def scan(im):
    i, (m, _, _) = im
    tok, c = booth_sessions[i % len(booth_sessions)]
    t0 = time.time()
    s, b = c.req("POST", "/api/v1/scans", token=tok, body={"member_code": m["member_code"]})
    return s, time.time() - t0, (b or {}).get("duplicate")

# One shared client per booth would serialize on the socket; give each worker
# its own connection by scanning through fresh clients in the hot loop.
def scan_fresh(im):
    i, (m, _, _) = im
    tok, _ = booth_sessions[i % len(booth_sessions)]
    c = Client()
    t0 = time.time()
    s, b = c.req("POST", "/api/v1/scans", token=tok, body={"member_code": m["member_code"]})
    return s, time.time() - t0, (b or {}).get("duplicate")

lat, statuses, results, wall = run_pool(list(enumerate(tokens)), scan_fresh, workers=200)
report("scan", lat, wall, statuses)
dupes = sum(1 for _, d in results if d)
check("every scan landed once, none duplicated",
      statuses.get(200, 0) == len(tokens) and dupes == 0, f"{statuses} dupes={dupes}")

# --------------------------------------------- 4b. networking check-in rush
print(f"\n== 4b. Networking check-in ({len(tokens)} people, {TABLES} tables of 8) ==")

def sit(im):
    i, (m, tok, c) = im
    table = (i % (len(tokens) // 8 + (1 if len(tokens) % 8 else 0))) + 1
    t0 = time.time()
    s, _ = c.req("POST", "/api/v1/networking/checkin", token=tok, body={"table_no": table})
    return s, time.time() - t0, table

lat, statuses, results, wall = run_pool(list(enumerate(tokens)), sit)
report("check-in", lat, wall, statuses)
check("everyone found a seat, no table oversold",
      statuses.get(200, 0) == len(tokens) and not any(k >= 500 for k in statuses),
      f"{statuses}")

print(f"\n{'='*40}\n{passed} passed, {failed} failed")
sys.exit(1 if failed else 0)
